nixos/lib/testing/network.nix

```diff
@@ -32,10 +32,19 @@ let
       # Automatically assign IP addresses to requested interfaces.
       assignIPs = lib.filter (i: i.assignIP) interfaces;
       ipInterfaces = forEach assignIPs (i:
-        nameValuePair i.name { ipv4.addresses =
-          [ { address = "192.168.${toString i.vlan}.${toString config.virtualisation.test.nodeNumber}";
+        nameValuePair i.name {
+          ipv4.addresses = [
+            {
+              address = "192.168.${toString i.vlan}.${toString config.virtualisation.test.nodeNumber}";
               prefixLength = 24;
-            }];
+            }
+          ];
+          ipv6.addresses = [
+            {
+              address = "2001:db8:${toString i.vlan}::${toString config.virtualisation.test.nodeNumber}";
```

Comment:
We should probably use a range from fc00::/7 instead of the documentation range

Reply:
Yeah, that prefix should be limited to examples in documentation.
Using a ULA is better, but it need to be picked [randomly](https://en.wikipedia.org/wiki/Unique_local_address#Definition) to avoid interfering with existing tests.

Reply:
Thanks for your review!
It seems the default configuration prefers ULAs over IPv4 as well, so I don't have a strong opinion on that.
Do you think I should change it right now, or is there a chance we might get another opinion?

Reply:
Using a GUA (here: the range reserved for documentation purposes) instead of a ULA has the benefit that IPv6 takes precedence over IPv4 when using `libc`'s `getaddrinfo`.

See [`gia.conf(5)`](https://man.archlinux.org/man/core/man-pages/gai.conf.5.en) and [`getaddrinfo(3)`](https://man.archlinux.org/man/getaddrinfo.3.en).

Reply:
@emilylange Are you sure this is actually the case? As per [RFC 3484](https://www.ietf.org/rfc/rfc3484.txt) section 10.3:

> The default policy table gives IPv6 addresses higher precedence than IPv4 addresses.  This means that applications will use IPv6 in preference to IPv4 when the two are equally suitable.

Both should be equally suitable as per section 3.2 since private IPv4 ranges are treated as site-local - the same as the ULA prefix - and an address of the same prefix is assigned to an interface or there is a route that matches the destination.

I investigated this since the described behavior in general seemed a bit odd to me and didn't match the default configuration of `gai.conf`. I believe that IPv6 will always be preferred if a suitable IPv6 address/route is available since the IPv6 mapped IPv4 address prefix `::ffff:0:0/96` has the lowest precedence:

```
      Prefix        Precedence Label
      ::1/128               50     0
      ::/0                  40     1
      2002::/16             30     2
      ::/96                 20     3
      ::ffff:0:0/96         10     4
```